resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -1197,6 +1206,34 @@ final DnsServerAddressStream newNameServerAddressStream(String hostname) {
         return dnsServerAddressStreamProvider.nameServerAddressStream(hostname);
     }
 
+    private static boolean isSpecialInvalid(final String hostname, final Promise<?> promise) {
+        if (INVALID.equalsIgnoreCase(hostname)) {
+            promise.setFailure(StacklessUnknownHostException.newInstance(
+                    "Special domain name \"invalid.\" always returns NXDOMAIN", DnsNameResolver.class,
+                    "doResolve(...)"));
+            return true;
+        }
+        return false;
+    }
+
+    private static final class StacklessUnknownHostException extends UnknownHostException {
```

Comment:
I wonder if we really need this as I don't expect that resolution of `invalid.` is on the hot-path. WDYT @idelpivnitskiy ?